cleverhans/attacks_tf.py

```diff
@@ -726,3 +726,135 @@ def compare(x, y):
+                        deepfool only consider nb_candidate classes when
+                        attacking (thus accelerate speed)
+    :param overshoot: A termination criterion to prevent vanishing updates
+    :param max_iter: Maximum number of iteration for DeepFool
+    :param clip_min: Minimum value for components of the example returned
+    :param clip_max: Maximum value for components of the example returned
+    :return: an adversarial sample
+    """
+    import copy
+
+    adv_x = copy.copy(sample)
+    # Initialize the loop variables
+    iteration = 0
+    current = utils_tf.model_argmax(sess, x, logits, adv_x, feed=feed)
+    if current.shape == ():
+        current = np.array([current])
+    w = np.squeeze(np.zeros(sample.shape[1:4]))  # same shape as original image
+    r_tot = np.zeros(sample.shape)
+    original = current  # use original label as the reference
+
+    _logger.debug("Starting DeepFool attack up to {} iterations".
+                  format(max_iter))
+    # Repeat this main loop until we have achieved misclassification
+    while (np.any(current == original) and iteration < max_iter):
+
+        if iteration % 5 == 0 and iteration > 0:
+            _logger.info("Attack result at iteration {} is {}".format(
+                iteration,
+                current))
+        gradients = sess.run(grads, feed_dict={x: adv_x})
+        predictions_val = sess.run(predictions, feed_dict={x: adv_x})
+        for idx in range(sample.shape[0]):
+            pert = np.inf
+            if current[idx] != original[idx]:
+                continue
+            for k in range(1, nb_candidate):
+                w_k = gradients[idx, k, ...] - gradients[idx, 0, ...]
+                f_k = predictions_val[idx, k] - predictions_val[idx, 0]
+                # adding value 0.00001 to prevent f_k = 0
+                pert_k = (abs(f_k) + 0.00001) / np.linalg.norm(w_k.flatten())
+                if pert_k < pert:
+                    pert = pert_k
+                    w = w_k
+            r_i = pert*w/np.linalg.norm(w)
+            r_tot[idx, ...] = r_tot[idx, ...] + r_i
+
+        adv_x = np.clip(r_tot + sample, clip_min, clip_max)
+        current = utils_tf.model_argmax(sess, x, logits, adv_x, feed=feed)
+        if current.shape == ():
+            current = np.array([current])
+        # Update loop variables
+        iteration = iteration + 1
+
+    # need more revision, including info like how many succeed
+    _logger.info("Attack result at iteration {} is {}".format(iteration,
+                 current))
+    _logger.info("{} out of {}".format(sum(current != original),
+                                       sample.shape[0]) +
+                 " becomes adversarial examples at iteration {}".format(
+                     iteration))
+    # need to clip this image into the given range
+    adv_x = np.clip((1+overshoot)*r_tot + sample, clip_min, clip_max)
+    return adv_x
+
+
+def gradient_graph(predictions, x, nb_candidate):
```

Comment:
Can you test this function separately from the attack class for DeepFool?

Reply:
The function gradient_graph is actually the same as the function jacobian_graph, with only difference at the output shape. We have deleted this function, and use jacobian_graph + reshape instead.